addons/product/models/product_template.py

```diff
@@ -489,6 +484,20 @@ def _name_search(self, name, args=None, operator='ilike', limit=100, name_get_ui
             '', args=[('id', 'in', list(searched_ids))],
             operator='ilike', limit=limit, name_get_uid=name_get_uid)
 
+    def _get_ids_of_tmpl_without_variant(self):
+        """
+        A template may not have any variant, or all variants of a template can be archived
+        The method returns the list of such templates
+        """
+        query = self.env['product.template']._where_calc([])
+        self.env['product.template']._apply_ir_rules(query, 'read')
+        query.add_join(('product_template', 'product_product', 'id', 'product_tmpl_id', 'product_product_rel'), implicit=False, outer=True, extra='{rhs}.active = TRUE')
+        query.where_clause.append('product_template__product_product_rel.id is NULL')
+        from_clause, where_clause, where_clause_params = query.get_sql()
+        query = "SELECT product_template.id FROM %s WHERE %s" % (from_clause, where_clause)
+        self.env.cr.execute(query, where_clause_params)
+        return [data[0] for data in self.env.cr.fetchall()]
```

Comment:
This is really some unreadable and highly unstable code.
It might be less performant, but can't you rely on simple search(es) with a correct domain ?
Custom query construction like this should be avoided, even more if you rely on hardcoded relation tables.

Can't you use the previous domain `[('product_variant_ids', '=', False)]` ?

Reply:
Nop, this domain needs to be improved: we need to include the templates whose all variants are archived. And we can't write such a query with the domains

As said to AWA, I agree the query construction is not so easily readable, but the query is simple. As an alternative, we could directly use the query and check the access on the result
```sql
select t.id 
from product_template t 
left join product_product p where t.id = p.product_tmpl_id AND product_product.active = TRUE
where p.id is NULL
```

Reply:
the issue with simple domain is that you need to get nearly all products out the the dabase in python, so something like:

````py
product_with_active_variant = self.env['product.template'].search([('product_variant_ids.active', '=', True])
self.env['product.template'].search([('id', 'not in', product_with_active_variant.ids)])
```
if there was a way to do that with without getting (nearly) all products of the database in python it would be a good alternative 
> even more if you rely on hardcoded relation tables.
there is no hardcoded relation table, this is the alias that we named in add_join parameters, resulting query is (in sudo to not have access right complicating it):

```sql
SELECT 
  product_template.id 
FROM 
  "product_template" 
  LEFT JOIN "product_product" as "product_template__product_product_rel" ON (
    "product_template"."id" = "product_template__product_product_rel"."product_tmpl_id" 
    AND product_template__product_product_rel.active = TRUE
  ) 
WHERE 
  (
    "product_template"."active" = true
  ) 
  AND product_template__product_product_rel.id is NULL
```

Reply:
> if there was a way to do that with without getting (nearly) all products of the database in python it would be a good alternative


 --> subqueries allow to delegate this to the SQL, but I think the ORM only supports this feature since 14+.

About the access check: since you are only providing the ids to the super name_search call, the super call will still apply the access rules anyway.

Reply:
> subqueries allow to delegate this to the SQL, but I think the ORM only supports this feature since 14+.

I have tried this and this seems to be working, only one query:

```py
self.env['product.template'].search([('id', 'not in', self.env['product.template']._search([('product_variant_ids.active', '=', True)]))])
```
so this would probably be better from 14.0, is this the syntax for subquery or is there a better way to write that?

Reply:
> About the access check: since you are only providing the ids to the super name_search call, the super call will still apply the access rules anyway.

so IMO (as well as awt) we could do simple query he gave (without using _where_calc, access_check, ...) in 13.0 and rely on name_search super for the filtering. And from 14.0 we could do the subquery fix.

Reply:
AFAIK, this is the correct syntax for subqueries.
Note that for 13.0, I just remembered there is the special `inselect` domain operator, but its use is restricted, I'm not 100% sure it can be used in this case.

Reply:
> Note that for 13.0, I just remembered there is the special inselect domain operator, but its use is restricted,

from cursory research, to use it you would need to evaluate the domain then add your leaf yourself with `internal=True` parameter, so (unless I am missing something) it doesn't seem this would be cleaner than current solutions or simple query.

Reply:
I knew it could be used for specific search methods defined for a specific field, but not how the restriction was applied in other cases.
Let's go for pure SQL here and subquery in 14+ then :ok_hand: